packages/transformers/webextension/src/WebExtensionTransformer.js

```diff
@@ -207,38 +221,97 @@ async function collectDependencies(
     const obj = parent[lastLoc];
     if (typeof obj == 'string')
       parent[lastLoc] = asset.addURLDependency(obj, {
-        needsStableName: true,
         loc: {
           filePath,
           ...getJSONSourceLocation(ptrs[location], 'value'),
         },
-        pipeline: path.extname(obj) == '.json' ? 'url' : undefined,
+        pipeline: path.extname(obj) == '.json' ? 'raw' : undefined,
       });
     else {
       for (const k of Object.keys(obj)) {
         obj[k] = asset.addURLDependency(obj[k], {
-          needsStableName: true,
           loc: {
             filePath,
             ...getJSONSourceLocation(ptrs[location + '/' + k], 'value'),
           },
-          pipeline: path.extname(obj[k]) == '.json' ? 'url' : undefined,
+          pipeline: path.extname(obj[k]) == '.json' ? 'raw' : undefined,
         });
       }
     }
   }
-  if (needRuntimeBG) {
-    if (!program.background) {
-      program.background = {};
+  if (isMV2) {
+    if (program.background?.page) {
+      program.background.page = asset.addURLDependency(
+        program.background.page,
+        {
+          loc: {
+            filePath,
+            ...getJSONSourceLocation(ptrs['/background/page'], 'value'),
+          },
+        },
+      );
+      if (needRuntimeBG) {
+        asset.meta.webextBGInsert = program.background.page;
+      }
     }
-    if (!program.background.scripts) {
-      program.background.scripts = [];
+    if (hot) {
+      // To enable HMR, we must override the CSP to allow 'unsafe-eval'
+      program.content_security_policy = cspPatchHMR(
+        program.content_security_policy,
+      );
+
+      if (needRuntimeBG && !program.background?.page) {
+        if (!program.background) {
+          program.background = {};
+        }
+        if (!program.background.scripts) {
+          program.background.scripts = [];
+        }
+        if (program.background.scripts.length == 0) {
+          program.background.scripts.push(
+            asset.addURLDependency('./runtime/default-bg.js', {
```

Comment:
I assume this empty script is needed for some reason?

Reply:
Yes, we need it to inject the auto-reloader for content-script page refresh. That's `autoreload-bg.js` in the runtime.